compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/OperatorPrecedence.java

```diff
@@ -63,6 +63,11 @@ public boolean isHigherThanOrEqual(OperatorPrecedence opPrecedence, boolean allo
                 return false;
             }
         }
-        return this.level <= opPrecedence.level;
+        boolean isHigherThanOrEqual = this.level <= opPrecedence.level;
+        // A query-action-or-expr or a query-action can have actions in certain clauses.
+        if (isHigherThanOrEqual && this == QUERY) {
+            return false;
+        }
+        return isHigherThanOrEqual;
```

Comment:
L68 is equivalent to `this == QUERY`?

```suggestion

        // A query-action-or-expr or a query-action can have actions in certain clauses.
        if (this == QUERY) {
            return false;
        }

        return this.level <= opPrecedence.level;;
```

Reply:
We can move the logic inside the `if (allowActions) { }` as this is only valid iff `allowActions = true`

Reply:
if we move this logic, the recovery doesn't work as expected for use of actions in query clauses that do not allow actions. For example
```ballerina
    _ = from var i in 1...3
        where obj->foo()
        select i;
```